[Docs] Add docs for reservations (#3895)
skypilot-org/skypilot: docs/source/reservations/reservations.rst

.. _reservation:

> Reservations & DWS

Reviewer: Why not simply Reservations / Reserved instances?

Author: DWS is not exactly a reservation, but more like a scheduler, as you don't know when the job will start in DWS. Also, I don't think many people think DWS as a reservation. Calling it `Reservation Instance` may make people miss the support of DWS. Another way to call this page is `Get High-End GPUs`. Wdyt?

Reviewer: `Reserved, Capacity Blocks, DWS`?

==================

Clouds are designed to be elastic, but due to the recent GPU shortages, many organizations have started to get reservations

> from cloud providers to lock in GPU capacity.

Reviewer: To discuss: it seems this belongs better to a blog rather than to a doc page.

Reviewer: Wdyt?

SkyPilot can be used to request resources from reservations and even combine them with on-demand/spot resources to fully
utilize the capacity in your cloud accounts.

> AWS capacity reservations & capacity block

Reviewer: ```suggestion
AWS capacity reservations & capacity blocks
```

------------------------------------------

> AWS capacity reservations and capacity block offers are a way to lock in a certain amount of compute capacity for a period of time. The latter is for high-end GPUs, such as A100s(P4d instances) and H100s(P5d instances), while the former is for all other instance types.
> Instead of committing to a 1-3 year long contract, you can get a capacity reservation and capacity block for as short as 1 second and 1 day, respectively.

Reviewer: ```suggestion
AWS **capacity reservations** and **capacity blocks** are ways to reserve a certain amount of compute capacity for a period of time. The latter is for high-end GPUs, such as A100s (P4d instances) and H100s (P5d instances), while the former is for all other instance types.
Instead of committing to a 1-3 year long contract, you can get a capacity reservation or capacity block for as short as 1 second or 1 day, respectively.
```

To request capacity reservations/blocks, see the official docs:

* `AWS Capacity Reservations <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-capacity-reservations.html>`_

> * `AWS Capacity Block Offers <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-capacity-blocks.html>`_

Reviewer: ```suggestion
* `AWS Capacity Blocks <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-capacity-blocks.html>`_
```

> Once you have got a reservation/block, you will get an id of your reservation, like ``cr-0123456789abcdefg``.
> 
> You can specify two fields in ``~/.sky/config.yaml``:

Reviewer: ```suggestion
Once you have successfully created a reservation/block, you will get an ID of the reservation/block, such as ``cr-0123456789abcdefg``.

To use the reservation/block, you can specify two fields in ``~/.sky/config.yaml``:
```

> * ``aws.prioritize_reservations``: whether to prioritize launching clusters from capacity reservations in any region/zone over on-demand/spot clusters. This is useful to fully utilize your reserved capacity created with *Instance eligibility: open*.
> * ``aws.specific_reservations``: a list of reservation ids that can be used by SkyPilot. This is useful if you have multiple capacity blocks or capacity reservations with *Instance eligibility: targeted* for different instance types in multiple regions/zones.

Reviewer: ```suggestion
* ``aws.prioritize_reservations``: whether to prioritize launching clusters from capacity reservations in any region/zone over on-demand/spot clusters. This is useful to fully utilize your reserved capacity created with ``Instance eligibility: open``.
* ``aws.specific_reservations``: a list of reservation IDs that can be used by SkyPilot. This is useful if you have multiple capacity reservations or blocks with ``Instance eligibility: targeted`` for different instance types in multiple regions/zones.
```

.. note::

>     If any of the fields are specified, please allow about 30 seconds for SkyPilot optimizer to retrieve the latest reservation/block status on all regions and zones from your AWS account.

Reviewer: Move this to after the snippet?

> 

Reviewer: ```suggestion

Example:

```

.. code-block:: yaml

    aws:
      prioritize_reservations: true
      specific_reservations:
        # 1x H100 capacity block in us-east-1
        - "cr-0123456789"
        # 2x A100 reservation in us-west-2
        - "cr-123456789a"
        # 2x M5a.16xlarge reservation in us-east-1
        - "cr-23456789ab"

> For more details of the configuration, see reference for :ref:`config-yaml`.

Reviewer: ```suggestion
For more details of the fields, see :ref:`config-yaml`.
```

.. _utilizing-reservations:

Utilizing Reservations
~~~~~~~~~~~~~~~~~~~~~~

> By specifying configurations above, SkyPilot will automatically figure out the best place to launch your clusters with available capacity in your reservations/blocks considered as zero-cost.

Reviewer: ```suggestion
By specifying the two fields above, SkyPilot will consider available capacity in your reservations/blocks as zero cost and automatically figure out the best place to launch your clusters.
```
Can we replace this to an even more intuitive sentence? Like "...will prioritize using any available reservation/block (i.e., consider them as zero cost) whenever you launch a cluster/job."

Author: Good point! Updated.

For example, if you are launching a cluster with the following SkyPilot YAML:

.. code-block:: yaml

    resources:
      cloud: aws
      accelerators: A100:8
    
    num_nodes: 4

> SkyPilot will utilize the capacity reservation as follows:

Reviewer: each bullet on a line to format

1. Find available capacity in your capacity block in ``us-west-2`` in the reservation ``cr-123456789a``. Let's say 1 A100 instance capacity is available.
2. SkyPilot will launch 4 nodes in ``us-west-2`` with 1 node coming from the reservation, and the rest 3 nodes from on-demand.
3. If it fails to find available A100 instances from on-demand, it will automatically :ref:`failover <auto-failover>` to other regions/zones.

> 

Author: Todo add a figure here

.. hint::

>     If you have a capacity block with a starting time in the future, you can use SkyPilot to launch a job on that capacity block as soon as the starting date comes, i.e. ``sky jobs launch --region us-east-1 --gpus H100:8 task.yaml``, so you don't have to wake up at 4:30am PDT to launch your job on a newly available capacity block.

Reviewer: "as soon as the starting date comes"

A bit confusing. Does this mean I cannot run this command on a prior date? Probably not; can we rephrase?

Author: Rephrased. PTAL.

>     If you have a capacity block with a starting time in the future, you can use SkyPilot to launch a job on that capacity block as soon as the starting date comes, i.e. ``sky jobs launch --region us-east-1 --gpus H100:8 task.yaml``, so you don't have to wake up at 4:30am PDT to launch your job on a newly available capacity block.

Reviewer: ```suggestion
    If you have a capacity block with a starting time in the future, you can use SkyPilot to launch a job on that capacity block at any time using ``sky jobs launch --region us-east-1 --gpus H100:8 task.yaml``. SkyPilot will automatically keep retrying to launch the job, and it will succeed as soon as the capacity block's start time hits. Namely, you don't have to wake up at 4:30am PDT to launch your job on a newly available capacity block.
```

GCP reservations
-----------------

GCP reservations are similar to AWS capacity reservations, where you can reserve a certain amount of compute capacity for any period of time.

> To get a reservation, see the `GCP official docs <https://cloud.google.com/compute/docs/instances/reservations-single-project>`__

Reviewer: ```suggestion
To get a reservation, see the `GCP official docs <https://cloud.google.com/compute/docs/instances/reservations-single-project>`__.
```

Like AWS, you can specify two fields in ``~/.sky/config.yaml``:

* ``gcp.prioritize_reservations``: whether to prioritize launching clusters from reservations in any region/zone over on-demand/spot clusters. This is useful to fully utilize your `automatically consumed reservations <https://cloud.google.com/compute/docs/instances/reservations-consume#consuming_instances_from_any_matching_reservation>`__.

> * ``gcp.specific_reservations``: a list of reservation ids that can be used by SkyPilot. This is useful if you have multiple `specific reservations <https://cloud.google.com/compute/docs/instances/reservations-consume#consuming_instances_from_a_specific_reservation>`__ for different instance types in multiple regions/zones.
> 
> .. code-block:: yaml

Reviewer: ```suggestion
* ``gcp.specific_reservations``: a list of reservation IDs that can be used by SkyPilot. This is useful if you have multiple `specific reservations <https://cloud.google.com/compute/docs/instances/reservations-consume#consuming_instances_from_a_specific_reservation>`__ for different instance types in multiple regions/zones.

Example:

.. code-block:: yaml
```

    gcp:
      prioritize_reservations: true
      specific_reservations:
        - projects/my-project/reservations/my-reservation1
        - projects/my-project/reservations/my-reservation2

SkyPilot will utilize the reservations similar to AWS reservations as described in :ref:`utilizing-reservations`.

GCP Dynamic Workload Scheduling (DWS)
-------------------------------------

> GCP `Dynamic Workload Scheduling (DWS) <https://cloud.google.com/blog/products/compute/introducing-dynamic-workload-scheduler>`__ is a resource management service that allows you submit a GPU capacity request, automatically provisions the requested resources when they are available and keeps the resources running for a specified duration.

Reviewer: s/Scheduling/Scheduler

> GCP Dynamic Workload Scheduling (DWS)
> -------------------------------------
> 
> GCP `Dynamic Workload Scheduling (DWS) <https://cloud.google.com/blog/products/compute/introducing-dynamic-workload-scheduler>`__ is a resource management service that allows you submit a GPU capacity request, automatically provisions the requested resources when they are available and keeps the resources running for a specified duration.

Reviewer: ```suggestion
GCP Dynamic Workload Scheduler (DWS)
-------------------------------------

GCP `Dynamic Workload Scheduler (DWS) <https://cloud.google.com/blog/products/compute/introducing-dynamic-workload-scheduler>`__ is a resource management service that (1) receives a GPU capacity request, (2) automatically provisions the requested resources when they become available, and (3) keeps the resources running for a specified duration.
```

> .. hint::

Reviewer: tip?

    It has been observed that using DWS can significantly increase the chance of getting a high-end GPU resource, such as A100s and H100s, compared to using on-demand or spot instances.

Using DWS for VMs
~~~~~~~~~~~~~~~~~

> SkyPilot allows you to launch a cluster with DWS by specifying the ``gcp.managed_instance_group`` field in ``~/.sky/config.yaml``:

Reviewer: ```suggestion
SkyPilot allows you to launch resources via DWS by specifying the ``gcp.managed_instance_group`` field in ``~/.sky/config.yaml``:
```

.. code-block:: yaml

    gcp:
      managed_instance_group:
        run_duration: 3600
        provision_timeout: 900

> The ``run_duration`` is the duration for a created instance to be kept alive (in seconds, required), and the ``provision_timeout`` is the timeout for provisioning an instance with DWS (in seconds, optional). If the timeout is reached without getting the requested resources, SkyPilot will automatically :ref:`failover <auto-failover>` to other clouds/regions/zones to get the resources.

Reviewer: ```suggestion
``run_duration`` is the duration for a created instance to be kept alive (in seconds, required). ``provision_timeout`` is the timeout for provisioning an instance with DWS (in seconds, optional). If the timeout is reached without requested resources being provisioned, SkyPilot will automatically :ref:`failover <auto-failover>` to other clouds/regions/zones to get the resources.
```

> See the :ref:`config-yaml` for more details.

Reviewer: ```suggestion
See :ref:`config-yaml` for more details.
```

In case you want to specify the DWS configuration for each job/cluster, you can also specify the configuration in the SkyPilot task YAML (see :ref:`here <task-yaml-experimental>`):

> .. code-block:: yaml

Reviewer: not rendered?

    experimental:
      config_overrides:
        gcp:
          managed_instance_group:
            run_duration: 3600
            provision_timeout: 900

    resources:
      cloud: gcp
      accelerators: A100:8
    
    num_nodes: 4
    
Using DWS on GKE with Kueue
~~~~~~~~~~~~~~~~~~~~~~~~~~

> DWS is also supported on Google Kubernetes Engine (GKE) with Kueue. To enable DWS on GKE, you need to set up your GKE cluster with Kueue and DWS, see the `GCP official docs <https://cloud.google.com/kubernetes-engine/docs/how-to/provisioningrequest>`

Reviewer: fix url

Author: I think this is the correct doc for DWS on GKE with Kueue?

To launch a SkyPilot cluster or job on GKE with DWS, you can specify the DWS configuration in the SkyPilot task YAML:

.. code-block:: yaml

    experimental:
      config_overrides:
        kubernetes:
          pod_config:
            metadata:
              annotations:
                provreq.kueue.x-k8s.io/maxRunDurationSeconds: "3600"
          provision_timeout: 900

    resourcse:
      cloud: kubernetes
      accelerators: A100:8
      labels:
        kueue.x-k8s.io/queue-name: dws-local-queue

> Here, ``kueue.x-k8s.io/queue-name`` is the name of the Kueue queue to submit your resource request to, and ``provreq.kueue.x-k8s.io/maxRunDurationSeconds`` is the maximum duration for a created instance to be kept alive (in seconds, required), and ``provision_timeout`` is the timeout for provisioning an instance with DWS (in seconds, optional). If the timeout is reached without getting the requested resources, SkyPilot will automatically :ref:`failover <auto-failover>` to other clouds/regions/zones to get the resources.

Author: @romilbhardwaj could you help check this part?

> Here, ``kueue.x-k8s.io/queue-name`` is the name of the Kueue queue to submit your resource request to, and ``provreq.kueue.x-k8s.io/maxRunDurationSeconds`` is the maximum duration for a created instance to be kept alive (in seconds, required), and ``provision_timeout`` is the timeout for provisioning an instance with DWS (in seconds, optional). If the timeout is reached without getting the requested resources, SkyPilot will automatically :ref:`failover <auto-failover>` to other clouds/regions/zones to get the resources.

Reviewer: split these 3 fields into 3 bullets for readability?

for last sentence, see comment above?

.. _reservation:

> Reserved, Capacity Blocks & DWS

Reviewer: ```suggestion
Reserved, Capacity Blocks, DWS
```

===================================

> Clouds are designed to be elastic, but due to the recent GPU shortages, many organizations have started to get reservations

Reviewer: ```suggestion
Clouds are designed to be elastic, but due to the recent GPU shortage, many organizations have started using reservations
```

from cloud providers to lock in GPU capacity.

SkyPilot can be used to request resources from reservations and even combine them with on-demand/spot resources to fully
utilize the capacity in your cloud accounts.

> .. image:: https://i.imgur.com/SbfnFDs.png

Reviewer: How about replacing that headline with "Get GPUs from reservations & other capacity types"?

Author: Sounds good. Updated.

3. :ref:`Automatically failover <auto-failover>` through these zones in increasing per-zone cost order until the requested resources are provisioned.

For example, if you are launching a cluster with the following SkyPilot YAML:

.. code-block:: yaml

    resources:
      cloud: aws
      accelerators: A100:8
    
    num_nodes: 2

SkyPilot will utilize the capacity reservation/block as follows:

1. Query reservations/blocks in ``us-east-2`` and ``us-west-2`` in reservation ``cr-123456789a`` and ``cr-23456789ab``, respectively. Assume the results are:

   - 1 A100 instance capacity is available in ``us-east-2``,
   - No available capacity in ``us-west-2``.
2. SkyPilot calculates the pricing for all zones as described above.  The result is ``us-east-2`` zones are cheaper than  all other zones, because the former's costs are 1 on-demand node's cost for 2 nodes (by satisfying 1 node using the reserved capacity).
3. SkyPilot will thus try to provision an on-demand A100 instance in ``us-east-2``. On unavailability, SkyPilot will continue to :ref:`automatically failover <auto-failover>` to other clouds/regions/zones for normal on-demand/spot instances.

.. hint::

>     If you have a capacity block with a starting time in the future, you can run ``sky jobs launch --region us-east-1 --gpus H100:8 task.yaml`` to let SkyPilot automatically wait until the starting time is reached. Namely, you don't have to wake up at 4:30am PDT to launch your job on a newly available capacity block.

Reviewer 2: Does this actually exist?

.. _reservation:

Reserved, Capacity Blocks, DWS
===================================

> With the recent GPU shortage, reservations from cloud providers for locking in GPU capacity have become one of the solutions to ensure the availability of GPU resources. Those reservations can be short-term (e.g., 1-30 days) capacity guarantees, or long-term (e.g., 1-3 years) contracts.

Reviewer: ```suggestion
With the recent GPU shortage, reservations from cloud providers have become a common way to ensure GPU availability for a specific duration. These reservations can be short-term (e.g., 1-30 days) capacity guarantees, or long-term (e.g., 1-3 years) contracts.
```

> This guide shows how SkyPilot can be used to request resources from reservations and even combine them with on-demand/spot resources to fully

Reviewer: ```suggestion
This guide shows how to use SkyPilot to request resources from reservations and even combine them with on-demand/spot resources to fully
```

SkyPilot can be used to request resources from reservations and even combine them with on-demand/spot resources to fully
utilize the capacity in your cloud accounts.

> .. image:: https://i.imgur.com/UY9eD1r.png

Reviewer: A bit confusing.

Author: Updated. Thanks!

      specific_reservations:
        # 1x H100 capacity block in us-east-1
        - "cr-0123456789"
        # 2x A100 reservation in us-east-2
        - "cr-123456789a"
        # 2x A100 reservation in us-west-2
        - "cr-23456789ab"
        # 2x M5a.16xlarge reservation in us-east-1
        - "cr-3456789abc"

For more details of the fields, see :ref:`config-yaml`.

.. note::

>     If any of the fields are specified, please allow about 30 seconds for SkyPilot optimizer to retrieve the latest reservation/block status on all regions and zones from your AWS account.

Reviewer: style: No need to use "Please" in tech docs / print messages.

"....SkyPilot optimizer may take around 30 seconds to ..."

.. note::

    If any of the fields are specified, SkyPilot optimizer may take around 30 seconds to retrieve the latest reservation/block status on all regions and zones from your AWS account.

.. _utilizing-reservations:

Utilizing Reservations
~~~~~~~~~~~~~~~~~~~~~~

By specifying configurations above, SkyPilot will prioritize using any available capacity in reservation/block (i.e., consider them as zero cost) whenever you launch a cluster/job.

> SkyPilot will act as follows:

Reviewer: ```suggestion
Specifically, SkyPilot's behavior is as follows:
```

> 1. Query reservations/blocks across each AWS regions and zones to find all non-occupied capacity.
> 2. Calculate per-zone pricing with non-occupied capacity as zero-cost and on-demand/spot price for the rest of resources (if any is needed to satisfy the requested resources).
> 3. :ref:`Automatically failover <auto-failover>` through different zones based on the per-zone pricing until the requested resources are satisfied.

Reviewer: ```suggestion
1. Query reservations/blocks across AWS regions and zones to find all available capacity. (If the task specifies specific regions or zones to use, only those are queried.)
2. For each zone, calculate its cost: any available reserved capacity is considered as zero cost, and if any on-demand/spot resource is needed to supplement the available reserved capacity to fully satisfy the request, their on-demand/spot price is included.
3. :ref:`Automatically failover <auto-failover>` through these zones in decreasing per-zone cost order until the requested resources are provisioned.
```

Author: Thanks! Applied the changes. `through these zones in decreasing per-zone cost order`. I changed this to `increasing` if I understand correctly

      cloud: aws
      accelerators: A100:8
    
    num_nodes: 2

> SkyPilot will utilize the capacity reservation as follows:

Reviewer: ```suggestion
SkyPilot will utilize the capacity reservation/block as follows:
```

> 1. Find available capacity in your capacity blocks in ``us-east-2`` and ``us-west-2`` in the reservation ``cr-123456789a`` and ``cr-23456789ab``, respectively. Let's say 1 A100 instance capacity is available in ``us-east-2`` but no available capacity in ``us-west-2`` (as shown in the figure above).

Reviewer: Not sure I fully understand this list based semantics (mentioned before). E.g., it seems step 3 needs to be expanded to say, failover to other places to look at cr-234....

Also is it guaranteed to look at cr-234... next?

Author: Updated the steps. It is not guaranteed to go to `cr-234` as it does not have any available capacity.

> 2. SkyPilot will first try to launch 2 nodes in ``us-east-2`` with 1 node coming from the reservation, and the rest 1 node as on-demand.

Reviewer: ```suggestion
2. SkyPilot will first try to launch 2 nodes in ``us-east-2`` with 1 node coming from the reservation, and the other node as on-demand.
```

> 3. If it fails to find available an on-demand A100 instance in ``us-east-2``, it will automatically :ref:`failover <auto-failover>` to other clouds/regions/zones.

Reviewer: ```suggestion
3. If it fails to find an available on-demand A100 instance in ``us-east-2``, it will automatically :ref:`failover <auto-failover>` to other clouds/regions/zones.
```

.. hint::

>     If you have a capacity block with a starting time in the future, you can run ``sky jobs launch --region us-east-1 --gpus H100:8 task.yaml`` and let SkyPilot to automatically wait until the starting time comes. Namely, you don't have to wake up at 4:30am PDT to launch your job on a newly available capacity block.

Reviewer: ```suggestion
    If you have a capacity block with a starting time in the future, you can run ``sky jobs launch --region us-east-1 --gpus H100:8 task.yaml`` to let SkyPilot automatically wait until the starting time is reached. Namely, you don't have to wake up at 4:30am PDT to launch your job on a newly available capacity block.
```

    num_nodes: 2

SkyPilot will utilize the capacity reservation/block as follows:

> 1. Query reservations/blocks in ``us-east-2`` and ``us-west-2`` in reservation ``cr-123456789a`` and ``cr-23456789ab``, respectively. Let's say:

Reviewer: ```suggestion
1. Query reservations/blocks in ``us-east-2`` and ``us-west-2`` in reservation ``cr-123456789a`` and ``cr-23456789ab``, respectively. Assume the results are:
```

The bullets don't render correctly

   - 1 A100 instance capacity is available in ``us-east-2``,
   - No available capacity in ``us-west-2``.

> 2. SkyPilot calculates the pricing for all zones and finds that zones in ``us-east-2`` are cheaper than other zones because it is considered as 1 on-demand price for 2 nodes.

Reviewer: ```suggestion
2. SkyPilot calculates the pricing for all zones as described above.  The result is ``us-east-2`` zones are cheaper than  all other zones, because the former's costs are 1 on-demand node's cost for 2 nodes (by satisfying 1 node using the reserved capacity).
```

> 3. If it fails to find an available on-demand A100 instance in ``us-east-2``, it will continue :ref:`automatically failover <auto-failover>` to other clouds/regions/zones for normal on-demand/spot instances.

Reviewer: ```suggestion
3. SkyPilot will thus try to provision an on-demand A100 instance in ``us-east-2``. On unavailability, SkyPilot will continue to :ref:`automatically failover <auto-failover>` to other clouds/regions/zones for normal on-demand/spot instances.
```

    num_nodes: 4
    
Using DWS on GKE with Kueue
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> DWS is also supported on Google Kubernetes Engine (GKE) with Kueue. To enable DWS on GKE, you need to set up your GKE cluster with Kueue and DWS, see the `GCP official docs <https://cloud.google.com/kubernetes-engine/docs/how-to/provisioningrequest>`__.

Reviewer: ```suggestion
DWS is also supported on Google Kubernetes Engine (GKE) with Kueue. To enable DWS on GKE, you need to set up your GKE cluster with Kueue and DWS; see the `GCP official docs <https://cloud.google.com/kubernetes-engine/docs/how-to/provisioningrequest>`__.
```
